compress.go

```diff
@@ -0,0 +1,244 @@
+	return &compressedReader{
+		buf:      buf,
+		bytesBuf: make([]byte, 0),
+		mc:       mc,
+	}
+}
+
+func NewCompressedWriter(connWriter io.Writer, mc *mysqlConn) *compressedWriter {
+	return &compressedWriter{
+		connWriter: connWriter,
+		mc:         mc,
+		header:     []byte{0, 0, 0, 0, 0, 0, 0},
+	}
+}
+
+func (cr *compressedReader) readNext(need int) ([]byte, error) {
+	for len(cr.bytesBuf) < need {
+		err := cr.uncompressPacket()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	data := cr.bytesBuf[:need]
+	cr.bytesBuf = cr.bytesBuf[need:]
+	return data, nil
+}
+
+func (cr *compressedReader) uncompressPacket() error {
+	header, err := cr.buf.readNext(7) // size of compressed header
+
+	if err != nil {
+		return err
+	}
+
+	// compressed header structure
+	comprLength := int(uint32(header[0]) | uint32(header[1])<<8 | uint32(header[2])<<16)
+	uncompressedLength := int(uint32(header[4]) | uint32(header[5])<<8 | uint32(header[6])<<16)
+	compressionSequence := uint8(header[3])
+
+	if compressionSequence != cr.mc.compressionSequence {
+		return ErrPktSync
+	}
+
+	cr.mc.compressionSequence++
+
+	comprData, err := cr.buf.readNext(comprLength)
+	if err != nil {
+		return err
+	}
+
+	// if payload is uncompressed, its length will be specified as zero, and its
+	// true length is contained in comprLength
+	if uncompressedLength == 0 {
+		cr.bytesBuf = append(cr.bytesBuf, comprData...)
+		return nil
+	}
+
+	// write comprData to a bytes.buffer, then read it using zlib into data
+	br := bytes.NewReader(comprData)
+
+	resetter, ok := cr.zr.(zlib.Resetter)
+
+	if ok {
+		err := resetter.Reset(br, []byte{})
+		if err != nil {
+			return err
+		}
+	} else {
+		cr.zr, err = zlib.NewReader(br)
+		if err != nil {
+			return err
+		}
+	}
+
+	defer cr.zr.Close()
+
+	//use existing capacity in bytesBuf if possible
+	offset := len(cr.bytesBuf)
+	if cap(cr.bytesBuf)-offset < uncompressedLength {
+		old := cr.bytesBuf
+		cr.bytesBuf = make([]byte, offset, offset+uncompressedLength)
+		copy(cr.bytesBuf, old)
+	}
+
+	data := cr.bytesBuf[offset : offset+uncompressedLength]
+
+	lenRead := 0
+
+	// http://grokbase.com/t/gg/golang-nuts/146y9ppn6b/go-nuts-stream-compression-with-compress-flate
+	for lenRead < uncompressedLength {
+		n, err := cr.zr.Read(data[lenRead:])
+		lenRead += n
+
+		if err == io.EOF {
+			if lenRead < uncompressedLength {
+				return io.ErrUnexpectedEOF
+			}
+			break
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	cr.bytesBuf = append(cr.bytesBuf, data...)
+
+	return nil
+}
+
+func (cw *compressedWriter) Write(data []byte) (int, error) {
+	// when asked to write an empty packet, do nothing
+	if len(data) == 0 {
+		return 0, nil
+	}
+	totalBytes := len(data)
+
+	length := len(data) - 4
+
+	maxPayloadLength := maxPacketSize - 4
+
+	for length >= maxPayloadLength {
+		// cut off a slice of size max payload length
+		dataSmall := data[:maxPayloadLength]
+		lenSmall := len(dataSmall)
+
+		var b bytes.Buffer
+		writer := zlib.NewWriter(&b)
```

Comment:
bytes.Buffer and zlib's writer are resettable.
At least, move them before loop.